akka-cluster-sharding/src/main/scala/akka/cluster/sharding/Shard.scala

```diff
@@ -388,4 +405,60 @@ private[akka] class PersistentShard(
     }
   }
 
+  private def retrieveEntityRecoveryConfig(): Config = {
+    val path = settings.tuningParameters.entityRecoveryStrategyConfigPath
+    if (path.isEmpty)
+      ConfigFactory.empty()
+    else
+      context.system.settings.config.getConfig(path)
+  }
+}
+
+abstract class EntityRecoveryStrategy {
+  import ShardRegion.EntityId
+  import akka.actor.ActorContext
+  def recoverEntities(context: ActorContext, entities: Set[EntityId])
```

Comment:
add explicit return type `: Unit`

Passing around `ActorContext` is a no-go. It's too easy to access it from wrong thread.
Why is it needed? I see sending `RestartEntities` to `self` and I see `scheduler`.

`RestartEntities` is something internal and could not be used by a user defined strategy anyway, so it might be the wrong api if user can't do what we can in those strategies.

The `EntityRecoveryStrategy` could take an optional `ActorSystem` constructor parameter to get access to scheduler and dispatcher things.

The API could perhaps be
```
def recoverEntities(entities: Set[EntityId]): Future[Set[EntityId]]
```

When the Future is completed we pipe the `RestartEntities` messages to the shard actor, i.e. the sending is not done from the strategy.

Another thought is that this whole pluggable strategy mechanism might be complete overkill. Instead we could provide 2 fixed implementations and implement them in the shard actor.